dm/master/bootstrap.go

```diff
@@ -0,0 +1,409 @@
+	"github.com/pingcap/dm/pkg/conn"
+	tcontext "github.com/pingcap/dm/pkg/context"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/terror"
+	"github.com/pingcap/dm/pkg/upgrade"
+	"github.com/pingcap/dm/pkg/v1dbschema"
+	"github.com/pingcap/dm/pkg/v1workermeta"
+)
+
+var (
+	// interval when waiting for the specified count of DM-worker instances become registered when importing from v1.0.x.
+	waitWorkerV1Interval = time.Second
+	// timeout when waiting for the specified count of DM-worker instances become registered when importing from v1.0.x.
+	waitWorkerV1Timeout = 5 * time.Minute
+)
+
+// bootstrap bootstraps the cluster, now including:
+// - upgrade the cluster from v1.0.x if needed.
+// - upgrade the cluster from a previous v2.0.x version to the current version.
+func (s *Server) bootstrap(ctx context.Context) error {
+	if s.cfg.V1SourcesPath != "" {
+		err := s.importFromV10x(ctx)
+		if err != nil {
+			return terror.ErrMasterFailToImportFromV10x.Delegate(err)
+		}
+	}
+
+	err := upgrade.TryUpgrade(s.etcdClient)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// importFromV10x tries to import/upgrade the cluster from v1.0.x.
+func (s *Server) importFromV10x(ctx context.Context) error {
+	// 1. check whether need to upgrade based on the cluster version.
+	preVer, _, err := upgrade.GetVersion(s.etcdClient)
+	if err != nil {
+		return err
+	} else if !preVer.NotSet() {
+		return nil // v2.0.x cluster, no need to import.
+	}
+
+	logger := log.L().WithFields(zap.String("op", "import from v1.0.x"))
+	tctx := tcontext.NewContext(ctx, logger)
+
+	// 2. collect source config files.
+	logger.Info("collecting source config files", zap.String("path", s.cfg.V1SourcesPath))
+	sourceCfgs, err := s.collectSourceConfigFilesV1Import(tctx)
+	if err != nil {
+		return err
+	}
+
+	// 3. wait for all DM-worker instances ready.
+	logger.Info("waiting for all DM-worker instances ready", zap.Int("count", len(sourceCfgs)))
+	err = s.waitWorkersReadyV1Import(tctx, sourceCfgs)
+	if err != nil {
+		return err
+	}
+
+	// 4. get subtasks config and stage from DM-worker instances.
+	logger.Info("getting subtask config and status from DM-worker")
+	subtaskCfgs, subtaskStages, err := s.getSubtaskCfgsStagesV1Import(tctx)
+	if err != nil {
+		return err
+	}
+
+	// 5. create sources.
+	logger.Info("add source config into cluster")
+	err = s.addSourcesV1Import(tctx, sourceCfgs)
+	if err != nil {
+		return err
+	}
+
+	// 6. upgrade v1.0.x downstream metadata table.
+	logger.Info("upgrading downstream metadata tables")
+	err = s.upgradeDBSchemaV1Import(tctx, subtaskCfgs)
+	if err != nil {
+		return err
+	}
+
+	// 7. create subtasks with the specified stage.
+	logger.Info("creating subtasks")
+	err = s.createSubtaskV1Import(tctx, subtaskCfgs, subtaskStages)
+	if err != nil {
+		return err
+	}
+
+	// 8. mark the upgrade operation as done.
+	logger.Info("marking upgrade from v1.0.x as done")
+	_, err = upgrade.PutVersion(s.etcdClient, upgrade.MinVersion)
```

Comment:
why not currentVersion 🤔

Reply:
if we importing v1.0.x in v2.0.1 (current), we may still need to do some operations (like create another metadata table/key) which should do in v2.0.0. If we put `currentVersion` here, the later upgrade operations will be skipped.